Domain Search Rewrite: Make sure Start flows behave correctly

zaguiini opened the pull request:
Part of DOMAINS-1674.

## Proposed Changes

Adds feature parity with production in all eligible `/start` flows, in order of importance:

1. `/start/launch-site?siteSlug=%s`
2. `/start/domain`
3. product flows (`/start/free`, `/start/premium`, `/start/business` etc.)
4. `/start/with-theme?theme=retrospect&theme_type=free&intervalType=yearly`
5. `/start/with-plugin?plugin=dokan-lite&billing_period=`
6. `/start/domain-for-gravatar`
7. `/start/onboarding-with-email`
8. `/start/onboarding-affiliate`

## Testing Instructions

Turn the `domain-search-rewrite` feature flag on and verify all of the flows behave like production: the ability to skip, to use an existing domain, the free domain discount, etc.

Don't worry about Tracks events as they'll be introduced in https://github.com/Automattic/wp-calypso/pull/105944.

zaguiini committed:
Allow using own domain in free flow

zaguiini committed:
Fix free domain submission

zaguiini committed:
Add back button

zaguiini committed:
Fix with-theme submission

zaguiini committed:
Pass currentSiteUrl to be used as a fallback to initial query

leonardost reviewed (commented):
I reviewed this and I'm not sure I'm doing it correctly 😅 I noticed some differences between this PR and production. For example:

Most of the flows are showing the "Exact match", "<TLD> is a common extension" notes in the features suggestions, but they aren't present in production. Also, the "Use a domain I own" link doesn't appear in this PR:

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/193c591e-008e-4036-acd8-3ac98930d78f" />

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/bfcdcf1a-82a8-484a-9b76-222c521a2a0c" />

Are these changes intentional?

I think the one that was most different was `/start/onboarding-with-email:

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/2a4c5c9f-ebf9-4233-8c51-9509b1da613f" />

Despite the UI changes and the omission of the "Use a domain I own" flow, I believe the behavior is the same with production.

zaguiini committed:
Don't offer free domain for first year in onboarding-with-email flow

zaguiini committed:
Lift query up

zaguiini committed:
Preserve professional email heading

zaguiini committed:
Make onboarding-with-email non-skippable

leonardost reviewed (commented):
Reviewed again and the new flows are more similar to the ones in prod 🙂 the behavior between this PR and prod is the same in all flows, but there are still some small UI differences. I've screenshotted each of the flows and pointed out those differences:

| Flow | Screenshot | Differences |
| --- | --- | --- |
| `/start/launch-site` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/6c0595b2-9d1c-4ff8-a173-e5d82aaa3620" /> | None ✅ |
| `/start/domain` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/00bd1390-16c7-400d-820c-6e3f3ae7b090" /> | "Use a domain I own" link and free domain banner only in the PR |
| `/start/domain` | <img width="956" height="1078" alt="Image" src="https://github.com/user-attachments/assets/6118d4df-0331-4917-b5ed-00234dc58ec1" /> | In the PR version, the "Use a domain I own" link appears over the mini cart - we should remove the link |
| `/start/free` |  <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/9b14e9e2-ca39-4700-a9ba-e14a045cec26" /> | Free domain banner only in prod |
| `/start/premium` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/c3e31135-1f1d-4c37-bc3b-9c8f940da36e" /> | None ✅ |
| `/start/with-theme` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/37215c18-fc51-4846-abd8-3361bbe7a3af" /> | PR uses "Free domain for 1st year" discount, prod doesn't |
| `/start/with-theme` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/8d08f257-2d80-4cb1-bf16-fac369e7de9f" /> | The "Use a domain I own" flow content has borders above and below that doesn't exist in prod |
| `/start/with-theme` | <img width="1920" height="1079" alt="Image" src="https://github.com/user-attachments/assets/86eed283-c73f-421f-bb86-0d9ed6389914" /> | Same as above |
| `/start/with-plugin` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/507bd6e4-7bd9-4b46-ae31-0db644965848" /> | WSOD in prod and this PR |
| `/start/domain-for-gravatar` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/4c817ab4-9eb3-45fd-99ea-b5ab8ebe9aa3" /> | None ✅ |
| `/start/onboarding-with-email` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/16c4229d-7ca5-470f-b4fe-a62e47a3b128" /> | Price rules are different |
| `/start/onbiarding-with-affiliate` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/e249e3a9-0998-4e3d-837f-a0d428787629" /> | None ✅ |
| `/start/onbiarding-with-affiliate` | <img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/a65af57c-6fc1-489d-96c8-c51e0901e549" /> | "Use a domain I own" has borders like in the `/start/with-theme` flow |

---

I noticed a weird behavior in the `/start/launch-site` domain search:

https://github.com/user-attachments/assets/43c5d84e-174a-488f-bc5b-8061762d64dd

I think this isn't related to this PR and the culprit seems to be the `hide_initial_query=yes` parameter. Removing it solves the issue with the query clearing on its own.

zaguiini committed:
Hide promo banner in domain flow

leonardost commented:
I've tested `/start/with-plugin?plugin=dokan-lite&billing_period=` and it works with the `billing_period` parameter now. The domain search page looks good, but if I click on "Use a domain I own", the flow thinks I've selected a domain:

<img width="1920" height="1080" alt="Screenshot 2025-09-29 at 16 58 27" src="https://github.com/user-attachments/assets/47474940-51d2-438a-8f4e-2f01a5bba734" />

<img width="1920" height="1080" alt="Screenshot 2025-09-29 at 17 04 57" src="https://github.com/user-attachments/assets/6fb85d84-ef6f-4006-a77a-a65689212800" />

https://github.com/user-attachments/assets/90b62e89-5d03-4b53-8f85-b02f1cf55d66

zaguiini committed:
Offer free domain for first year in onboarding with email flow

zaguiini committed:
Remove action buttons z-index

zaguiini committed:
Remove use my domain border

zaguiini committed:
Fix use my domain navigation

zaguiini commented:
Thanks for the fantastic review, @leonardost!

I think I addressed most of concerns, notably excepting:
- `/start/domain` should have the use my domain link at the top-right because it's possible to transfer a domain from that flow. Prod not having it is a bug;
- `/start/free` shouldn't have the free plan banner and that's a bug in prod. We're not offering plans in that flow;
- `/start/with-theme` I think having the free for a year discount is actually correct as we're showing the plans step as the user progresses. It's a production bug apparently. The same for `onboarding-with-email`.

leonardost reviewed (approved):
Thanks for all the work on this, @zaguiini!

> /start/domain should have the use my domain link at the top-right because it's possible to transfer a domain from that flow. Prod not having it is a bug;
> /start/free shouldn't have the free plan banner and that's a bug in prod. We're not offering plans in that flow;
> /start/with-theme I think having the free for a year discount is actually correct as we're showing the plans step as the user progresses. It's a production bug apparently. The same for onboarding-with-email.

With these considerations, I believe everything LGTM!

There's one very minor thing, there's a border in the second step of the "Use a domain I own" flow:

<img width="1920" height="1080" alt="Screenshot 2025-09-29 at 18 17 17" src="https://github.com/user-attachments/assets/aafff336-a78a-4a4d-83dd-fa96319252a3" />

It shows in all the flows except for `/start/launch-site`.

zaguiini committed:
Remove border from use my domain child steps
